executor/coprocessor.go

```diff
@@ -0,0 +1,187 @@
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/planner/core"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/timeutil"
+	"github.com/pingcap/tipb/go-tipb"
+)
+
+// CoprocessorDAGHandler uses to handle cop dag request.
+type CoprocessorDAGHandler struct {
+	sctx    sessionctx.Context
+	resp    *coprocessor.Response
+	selResp *tipb.SelectResponse
+	dagReq  *tipb.DAGRequest
+}
+
+// NewCoprocessorDAGHandler creates a new CoprocessorDAGHandler.
+func NewCoprocessorDAGHandler(sctx sessionctx.Context, resp *coprocessor.Response) *CoprocessorDAGHandler {
+	return &CoprocessorDAGHandler{
+		sctx:    sctx,
+		resp:    resp,
+		selResp: &tipb.SelectResponse{},
+	}
+}
+
+// HandleCopDAGRequest handles the coprocessor request.
+func (h *CoprocessorDAGHandler) HandleCopDAGRequest(ctx context.Context, req *coprocessor.Request) *coprocessor.Response {
+	e, err := h.buildDAGExecutor(req)
+	if err != nil {
+		return h.buildResp(err)
+	}
+
+	err = e.Open(ctx)
+	if err != nil {
+		return h.buildResp(err)
+	}
+
+	chk := newFirstChunk(e)
+	tps := e.base().retFieldTypes
+	for {
+		chk.Reset()
+		err = Next(ctx, e, chk)
+		if err != nil {
+			break
+		}
+		if chk.NumRows() == 0 {
+			break
+		}
+		err = h.fillUpData4SelectResponse(chk, tps)
+		if err != nil {
+			break
+		}
+	}
+	return h.buildResp(err)
+}
+
+func (h *CoprocessorDAGHandler) buildDAGExecutor(req *coprocessor.Request) (Executor, error) {
+	if req.GetTp() != kv.ReqTypeDAG {
+		return nil, errors.Errorf("unsupported request type %d", req.GetTp())
+	}
+	dagReq := new(tipb.DAGRequest)
+	err := proto.Unmarshal(req.Data, dagReq)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	h.sctx.GetSessionVars().StmtCtx.SetFlagsFromPBFlag(dagReq.Flags)
+	h.sctx.GetSessionVars().StmtCtx.TimeZone, err = timeutil.ConstructTimeZone(dagReq.TimeZoneName, int(dagReq.TimeZoneOffset))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	h.dagReq = dagReq
+	e, err := h.buildDAG(dagReq.Executors)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return e, nil
+}
+
+func (h *CoprocessorDAGHandler) buildDAG(executors []*tipb.Executor) (Executor, error) {
+	is := h.sctx.GetSessionVars().TxnCtx.InfoSchema.(infoschema.InfoSchema)
+	bp := core.NewPBPlanBuilder(h.sctx, is)
+	plan, err := bp.BuildPhysicalPlanFromPB(executors)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	b := newExecutorBuilder(h.sctx, is)
+	return b.build(plan), nil
+}
+
+func (h *CoprocessorDAGHandler) fillUpData4SelectResponse(chk *chunk.Chunk, tps []*types.FieldType) error {
+	var err error
+	switch h.dagReq.EncodeType {
+	case tipb.EncodeType_TypeDefault:
+		err = h.encodeDefault(chk, tps)
+	case tipb.EncodeType_TypeChunk:
+		err = h.encodeChunk(chk, tps)
+	default:
+		return errors.Errorf("unknown dag encode type, %v", h.dagReq.EncodeType)
+	}
+	h.selResp.EncodeType = h.dagReq.EncodeType
```

Comment:
I think `h.selResp.EncodeType = h.dagReq.EncodeType` can be moved to `buildResponse` instead of executing every time.

Reply:
done.